enhancements/windows-containers/disconnected-support.md

```diff
@@ -0,0 +1,372 @@
+nodes.
+QE should cover disconnected environments on all supported platforms when validating this feature.
+
+### User Stories
+
+Story 1: Set `config_path` in containerd_conf.toml
+Story 2: WMCO generates hosts.toml files
+Story 3: WMCO copies over generated hosts.toml files to each Windows instance
+
+User stories can also be found within the JIRA epic: [WINC-936](https://issues.redhat.com/browse/WINC-936)
+
+### Workflow Description and Variations
+
+**cluster creator** is a human user responsible for deploying a cluster.
+**cluster administrator** is a human user responsible for managing cluster settings including network egress policies.
+
+There are 4 different workflows that affect the disconnected use case.
+1. A cluster creator brings up a new disconnected cluster and configures mirror registry settings at install time
+2. A cluster administrator introduces new mirror registry settings or changes exisiting ones during runtime
+   This would occur through the creation, deletion, or update IDMS/ITMS resources.
+3. A cluster administrator transitions a connected cluster *with existing mirror registry settings* to a disconnected cluster
+4. A cluster administrator transitions a connected cluster *without existing mirror registry settings* to a disconnected cluster
+
+
+In cases 1, 2 & 3, Windows nodes can be joined to the cluster after altering registry settings,
+which would result in WMCO applying mirror settings during initial node configuration. 
+
+In case 2, for any existing Windows nodes, WMCO will react by updating each instance's state to consume the new registry config.
+
+In case 3, existing Windows nodes will not be re-configured. Existing registry settings will stay applied.
+
+In case 4, WMCO will take no action as there are no settings to configure. Windows nodes will not be able to pull images
+until mirror registry settings are configured by the administrator.
+
+### Risks and Mitigations
+
+If node cleanup fails, instances may retain their containerd configuration. This could be a concern for BYOH instances
+particularly. An event should be generated to alert cluster admins manual cleanup is required to remove WMCO-created files.
+
+Also, WMCO has no way of detecting local changes to the containerd registry config directory on Windows nodes.
+The changes will remain in effect until a new IDMS/ITMS change is made in the cluster or the node is reconfigured.
+
+Another concern is if a user's mirror configuration has overlapping settings/clashes. As on the Linux side, it is upon
+the user to ensure there is no clash between mirrors across their IDMS and ITMS resources.
+
+<!--
+Containerd may not support all the fields available in IDMS/ITMS resources.
+Unsupported fields will be ignored and will not impact containerd registry config/default behavior.
+-->
+
+### Drawbacks
+
+The major drawback with this design is a scaling issue. WMCO will have to SSH into each node to update mirror registry
```

Comment:
Don't we have a per node entity that can watch and apply this concurrently?

Reply:
Yes we do, the Windows Instance Config Daemon (WICD) runs as a service on each Windows node. The previous draft of the enhancement actually followed this design
- WMCO consumes cluster mirror settings to generate the containerd config files
- WMCO publishes the files to WICD (through a configmap that WICD watches)
- WICD consumes configmap data to create files on Windows node, updating mirror settings

This would require us to create a mechanism similar to how MCO stores files in MachineConfigs. The team had a meeting around this and concluded (with Aravindh) that this is too complex of an approach for the disconnected MVP. We decided to implement the simpler solution, and then take on a future epic to introduce a `MachineConfigSpec`-like object.

TLDR: we heavily considered it but eventually decided against it

Reply:
How many maximum number of nodes can we support with the current approach? That's something we would want documented.

Reply:
I don't think we can put a number on it, but the time it takes to update mirror settings will scale linearly with the # of Windows nodes in the cluster in the worst case. IMO this won't affect the # nodes we can support from a technical standpoint. The limit more so depends on a user's tolerance for longer re-configuration times and how often do they change their mirror settings.

One thing to note is that we are introducing a new controller that will only handle events related to image mirroring. This means that we won't have to enter into a full node re-configuration when cluster mirror settings change.

Reply:
I think we can proceed with clear documentation covering this and then if this becomes an issue for adoption, we should get the other work prioritized to make it scale.